src/client/Client.h

```diff
@@ -1334,8 +1334,7 @@ class Client : public Dispatcher, public md_config_obs_t {
     // calls to complete. All the handling though is in C_Read_Finisher.
```

Comment:
:scream: This is such a no-no. @kotreshhr why was it done this way?

Reply:
I'm not 100% sure if I understand your comment. Do you mean having the context callback invoked multiple times is the issue here?

Reply:
> I'm not 100% sure if I understand your comment. Do you mean having the context callback invoked multiple times is the issue here?

Yes! Contexts should only ever be finished once.

Reply:
Really? I've surely seen code that overrides complete() and conditionally deletes `this`. I think it was in librbd.

Reply:
Well, I'd say that's a weirdness we should not continue elsewhere. It's just too confusing. (I think the `Context` interface has acquired so many warts that it's already stupifying. Let's not add to it.)

Reply:
I wish I could remember all the gyrations I went through with this code...

Some crap got in because some things changed under me while I was developing it and I probably didn't always correctly account for the changes.

I also remember hitting some derived class of Context that reversed the meaning of finish and complete which threw me for quite a loop. And I think as the implementation changed I eventually did not derive from THAT class, but may not have corrected all the assumptions about finish and complete...

At this point @dparmar18 is the expert on this code, having fixed several of my goofs...

Reply:
I think the potential for multiple finishes comes from C_Read_Sync_NonBlocking, it can retry.

But I think something is missing to actually connect all the dots there...